File: lib/java/src/main/java/io/github/telepact/internal/schema/DocumentLocators.java
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package io.github.telepact.internal.schema;

import static io.github.telepact.internal.schema.GetPathDocumentCoordinatesPseudoJson.getPathDocumentCoordinatesPseudoJson;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class DocumentLocators {
    @FunctionalInterface
    public interface DocumentLocator {
        Map<String, Object> locate(List<Object> path);
    }

    public static class SchemaDocumentMap extends HashMap<String, String> {
        public final Map<String, DocumentLocator> documentLocators = new HashMap<>();

        public SchemaDocumentMap() {
            super();
        }

        public SchemaDocumentMap(Map<String, String> other) {
            super(other);
            if (other instanceof SchemaDocumentMap schemaDocumentMap) {
                this.documentLocators.putAll(schemaDocumentMap.documentLocators);
            }
        }
    }

    public static SchemaDocumentMap copyDocumentLocators(Map<String, String> source, Map<String, String> destination) {
        final var mapped = destination instanceof SchemaDocumentMap schemaDocumentMap
                ? schemaDocumentMap
                : new SchemaDocumentMap(destination);
        if (source instanceof SchemaDocumentMap schemaSource) {
            mapped.documentLocators.putAll(schemaSource.documentLocators);
        }
        return mapped;
    }

    public static Map<String, Object> resolveDocumentCoordinates(
            List<Object> path,
            String documentName,
            Map<String, String> documentNamesToJson) {
        if (documentNamesToJson instanceof SchemaDocumentMap schemaDocumentMap) {
            final var locator = schemaDocumentMap.documentLocators.get(documentName);
            if (locator != null) {
                return locator.locate(path);
            }
        }
        return getPathDocumentCoordinatesPseudoJson(path, documentNamesToJson.get(documentName));
    }
}

File: lib/py/telepact/internal/schema/DocumentLocators.py
#|
#|  Copyright The Telepact Authors
#|  SPDX-License-Identifier: Apache-2.0
#|

from typing import Callable


DocumentLocator = Callable[[list[object]], dict[str, object]]


class SchemaDocumentMap(dict[str, str]):
    document_locators: dict[str, DocumentLocator]

    def __init__(self, *args: object, **kwargs: object):
        super().__init__(*args, **kwargs)
        self.document_locators = {}


def ensure_schema_document_map(document_names_to_json: dict[str, str]) -> SchemaDocumentMap:
    if isinstance(document_names_to_json, SchemaDocumentMap):
        return document_names_to_json

    mapped = SchemaDocumentMap(document_names_to_json)
    locators = getattr(document_names_to_json, 'document_locators', None)
    if isinstance(locators, dict):
        mapped.document_locators = dict(locators)
    return mapped


def set_document_locators(document_names_to_json: dict[str, str], document_locators: dict[str, DocumentLocator]) -> None:
    mapped = ensure_schema_document_map(document_names_to_json)
    mapped.document_locators = dict(document_locators)


def get_document_locators(document_names_to_json: dict[str, str]) -> dict[str, DocumentLocator]:
    mapped = ensure_schema_document_map(document_names_to_json)
    return mapped.document_locators


def copy_document_locators(src: dict[str, str], dst: dict[str, str]) -> SchemaDocumentMap:
    mapped = ensure_schema_document_map(dst)
    mapped.document_locators = dict(getattr(src, 'document_locators', {}))
    return mapped


def resolve_document_coordinates(path: list[object], document_name: str, document_names_to_json: dict[str, str]) -> dict[str, object]:
    from .GetPathDocumentCoordinatesPseudoJson import get_path_document_coordinates_pseudo_json

    locator = getattr(document_names_to_json, 'document_locators', {}).get(document_name)
    if locator is not None:
        return locator(path)

    return get_path_document_coordinates_pseudo_json(path, document_names_to_json.get(document_name, '[]'))
